x/application/keeper/msg_server_undelegate_from_gateway_test.go

```diff
@@ -272,3 +281,225 @@ func TestMsgServer_UndelegateFromGateway_SuccessfullyUndelegateFromUnstakedGatew
 	require.Equal(t, appAddr, foundApp.Address)
 	require.Equal(t, 0, len(foundApp.DelegateeGatewayAddresses))
 }
+
+func TestMsgServer_UndelegateFromGateway_DelegationIsActiveUntilNextSession(t *testing.T) {
+	k, ctx := keepertest.ApplicationKeeper(t)
+	srv := keeper.NewMsgServerImpl(k)
+
+	undelegationHeight := int64(1)
+	sdkCtx, app, gatewayAddrs := delegateThenUndelegate(ctx, t, srv, k, undelegationHeight)
+
+	gatewayAddrToUndelegate := gatewayAddrs[0]
+	delegatedGatewayAddr := gatewayAddrs[1]
+
+	// Increment the block height without moving to the next session, then run the
+	// pruning undelegations logic.
+	sdkCtx = sdkCtx.WithBlockHeight(undelegationHeight + 1)
+	k.EndBlockerPruneAppToGatewayPendingUndelegation(sdkCtx)
+
+	// Get the updated application state after pruning logic is run.
+	app, isAppFound := k.GetApplication(sdkCtx, app.Address)
+	require.True(t, isAppFound)
+	require.NotNil(t, app)
+
+	// Verify that the gateway was removed from the application's delegatee gateway addresses.
+	require.NotContains(t, app.DelegateeGatewayAddresses, gatewayAddrToUndelegate)
+
+	// Verify that the gateway is added to the pending undelegation list with the
+	// right sessionEndHeight as the map key.
+	sessionEndHeight := uint64(sessionkeeper.GetSessionEndBlockHeight(undelegationHeight))
+	require.Contains(t,
+		app.PendingUndelegations[sessionEndHeight].GatewayAddresses,
+		gatewayAddrToUndelegate,
+	)
+
+	// Verify that the other gateways are still delegated to the application.
+	require.Contains(t, app.DelegateeGatewayAddresses, delegatedGatewayAddr)
+
+	// Verify that the reconstructed delegatee gateway list include the undelegated gateway.
+	gatewayAddresses := rings.GetRingAddressesAtBlock(&app, sdkCtx.BlockHeight())
+	require.Contains(t, gatewayAddresses, gatewayAddrToUndelegate)
+
+	// Increment the block height to the next session and  run the pruning
+	// undelegations logic again.
+	nextSessionStartHeight := int64(sessionEndHeight + 1)
+	sdkCtx = sdkCtx.WithBlockHeight(nextSessionStartHeight)
+	k.EndBlockerPruneAppToGatewayPendingUndelegation(sdkCtx)
+
+	// Get the updated application state.
+	app, isAppFound = k.GetApplication(sdkCtx, app.Address)
+	require.True(t, isAppFound)
+	require.NotNil(t, app)
+
+	// Verify that when queried for the next session the reconstructed delegatee
+	// gateway list does include the undelegated gateway.
+	nextSessionGatewayAddresses := rings.GetRingAddressesAtBlock(&app, nextSessionStartHeight)
+	require.NotContains(t, nextSessionGatewayAddresses, gatewayAddrToUndelegate)
```

Comment:
Can you please add a few more steps to make sure the same thing applies when you take the grace period into consideration?